From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MichaelSimons <msimons@microsoft.com>
Date: Mon, 6 Dec 2021 18:39:25 +0000
Subject: [PATCH] WebHost delegate nullability

Addresses the following errors:

  src/Hosting/Hosting/src/Internal/WebHost.cs(83,56): error CS8621: Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'Func<IServiceProvider, IHostApplicationLifetime?>' (possibly because of nullability attributes).
  src/Hosting/Hosting/src/Internal/WebHost.cs(86,56): error CS8621: Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'Func<IServiceProvider, IApplicationLifetime?>' (possibly because of nullability attributes).
  src/Hosting/Hosting/src/Internal/WebHost.cs(88,56): error CS8621: Nullability of reference types in return type of 'lambda expression' doesn't match the target delegate 'Func<IServiceProvider, IApplicationLifetime?>' (possibly because of nullability attributes).

Cherry-pick of WebHost only changes made under https://github.com/dotnet/aspnetcore/commit/13a460820e90c8f3e90eb4fdbbb3de671389bf0c#diff-6ef451f2ae3e04d73dde362bdf9de6b101aef95e3a4f5190e5c96967d683aef9
---
 src/Hosting/Hosting/src/Internal/WebHost.cs | 14 ++++++--------
 1 file changed, 6 insertions(+), 8 deletions(-)

diff --git a/src/Hosting/Hosting/src/Internal/WebHost.cs b/src/Hosting/Hosting/src/Internal/WebHost.cs
index ca52bd1f40..643d03e8af 100644
--- a/src/Hosting/Hosting/src/Internal/WebHost.cs
+++ b/src/Hosting/Hosting/src/Internal/WebHost.cs
@@ -79,16 +79,14 @@ namespace Microsoft.AspNetCore.Hosting
             _hostingServiceProvider = hostingServiceProvider;
             _applicationServiceCollection.AddSingleton<ApplicationLifetime>();
             // There's no way to to register multiple service types per definition. See https://github.com/aspnet/DependencyInjection/issues/360
-#pragma warning disable CS8634 // The type cannot be used as type parameter in the generic type or method. Nullability of type argument doesn't match 'class' constraint.
-            _applicationServiceCollection.AddSingleton(services
-                => services.GetService<ApplicationLifetime>()! as IHostApplicationLifetime);
+             _applicationServiceCollection.AddSingleton<IHostApplicationLifetime>(services
+                => services.GetService<ApplicationLifetime>()!);
 #pragma warning disable CS0618 // Type or member is obsolete
-            _applicationServiceCollection.AddSingleton(services
-                => services.GetService<ApplicationLifetime>()! as AspNetCore.Hosting.IApplicationLifetime);
-            _applicationServiceCollection.AddSingleton(services
-                => services.GetService<ApplicationLifetime>()! as Extensions.Hosting.IApplicationLifetime);
+            _applicationServiceCollection.AddSingleton<AspNetCore.Hosting.IApplicationLifetime>(services
+                => services.GetService<ApplicationLifetime>()!);
+            _applicationServiceCollection.AddSingleton<Extensions.Hosting.IApplicationLifetime>(services
+                => services.GetService<ApplicationLifetime>()!);
 #pragma warning restore CS0618 // Type or member is obsolete
-#pragma warning restore CS8634 // The type cannot be used as type parameter in the generic type or method. Nullability of type argument doesn't match 'class' constraint.
             _applicationServiceCollection.AddSingleton<HostedServiceExecutor>();
         }
 
